django/db/models/sql/query.py

```diff
@@ -1605,7 +1605,11 @@ def build_filtered_relation_q(
 
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
-        lookups = dict(get_children_from_q(filtered_relation.condition))
+        lookups = dict(
+            l
+            for l in get_children_from_q(filtered_relation.condition)
+            if isinstance(l, tuple)
```

Comment:
I think this is only _masking_ the crash by dropping the false criteria

If it was generating a join along the lines of `LEFT JOIN book AS the_book ON (book.author_id = author.id AND false)`, as should be expected in this case, then `the_book.id` should always be `NULL` as `(book.author_id = author.id AND false)` will always be false.

Your test demonstrates that the criteria is dropped and produces wrong results.